Cognite.Simulator.Utils/ConcurrentTaskManager.cs

```diff
@@ -0,0 +1,98 @@
+
+        /// <summary>
+        /// Creates a new <see cref="ConcurrentTaskManager{TKey, TResult}"/>.
+        /// </summary>
+        /// <param name="maxConcurrentTasks">Maximum number of tasks allowed to run concurrently.</param>
+        public ConcurrentTaskManager(int maxConcurrentTasks = 5)
+        {
+            if (maxConcurrentTasks <= 0)
+                throw new ArgumentOutOfRangeException(nameof(maxConcurrentTasks), "maxConcurrentTasks must be greater than 0.");
+            _semaphore = new SemaphoreSlim(maxConcurrentTasks, maxConcurrentTasks);
+            _ongoingTasks = new ConcurrentDictionary<TKey, (Task<TResult>, CancellationTokenSource)>();
+        }
+
+        /// <summary>
+        /// Executes a task for a given key, ensuring deduplication and respecting the concurrency limit.
+        /// </summary>
+        /// <param name="key">Unique key identifying the task.</param>
+        /// <param name="taskFactory">Function creating the task. It receives the <paramref name="token"/> so it can honour cancellation.</param>
+        /// <param name="token">Cancellation token.</param>
+        /// <returns>The task result.</returns>
+        /// <exception cref="ObjectDisposedException">Thrown if the manager has been disposed.</exception>
+        public Task<TResult> ExecuteAsync(TKey key, Func<CancellationToken, Task<TResult>> taskFactory, CancellationToken token = default)
+        {
+            if (key == null) throw new ArgumentNullException(nameof(key));
+            if (taskFactory == null) throw new ArgumentNullException(nameof(taskFactory));
+            if (_disposed) throw new ObjectDisposedException(nameof(ConcurrentTaskManager<TKey, TResult>));
+
+            var cts = CancellationTokenSource.CreateLinkedTokenSource(token);
+            Task<TResult> task = null;
+
+            var entry = _ongoingTasks.GetOrAdd(key, k =>
+            {
+                task = CreateManagedTask(k, taskFactory, cts.Token);
+                return (task, cts);
+            });
+
+            if (task == null)
+            {
+                cts.Dispose();
+                return entry.Task;
+            }
+
+            return task;
+        }
+
+        private Task<TResult> CreateManagedTask(TKey key, Func<CancellationToken, Task<TResult>> taskFactory, CancellationToken token)
+        {
+            return Task.Run(async () =>
+            {
+                await _semaphore.WaitAsync(token).ConfigureAwait(false);
+                try
+                {
+                    return await taskFactory(token).ConfigureAwait(false);
+                }
+                finally
+                {
+                    try { _semaphore.Release(); } catch (ObjectDisposedException) { /* Ignore: Semaphore may be disposed during shutdown */ }
+                    if (_ongoingTasks.TryRemove(key, out var removed))
+                        removed.Cts?.Dispose();
+                }
+            }, token);
+        }
```

Comment:
```suggestion
        private async Task<TResult> CreateManagedTask(TKey key, Func<CancellationToken, Task<TResult>> taskFactory, CancellationToken token)
        {
            await _semaphore.WaitAsync(token).ConfigureAwait(false);
            try
            {
                return await taskFactory(token).ConfigureAwait(false);
            }
            finally
            {
                try { _semaphore.Release(); } catch (ObjectDisposedException) { /* Ignore: Semaphore may be disposed during shutdown */ }
                if (_ongoingTasks.TryRemove(key, out var removed))
                    removed.Cts?.Dispose();
            }
        }
```
this simplifies the code (add WaitAsync, remove Task.Run)